### ISSUE TITLE
TPOT 2 vs TPOT 1 huge runtime difference

### ISSUE THREAD

**Issue by christofplie** [EpistasisLab/tpot#1388]
**Context**
I am conducting a scientific experiment on AutoML reproducibility across five TPOT versions (0.11.6, 0.11.7, 0.12.1, 0.12.2, and 1.0.0). I am running these on a local machine (no cloud infrastructure) using Docker containers (Linux).

**The Issue**
I have observed a significant runtime regression in v1.0.0 compared to all previous versions, despite using the same dataset and hyperparameter configuration.

Dataset: Regression task, 638 samples, 13 features.

Dependencies: xgboost is installed and available in all environments.

Hardware: Local PC (Docker), identical resources allocated for all runs.

**Observed Results**
TPOT v0.11.6 / v0.11.7 (Python 3.8): Runtime ~4-5 minutes.

TPOT v0.12.1 / v0.12.2 (Python 3.10): Runtime ~4-5 minutes.

TPOT v1.0.0 (Python 3.10): Runtime ~30 minutes.

**Configuration**
The code logic is identical across versions (with v1.0.0 wrapped in if __name__ == '__main__': to support the Dask backend).
Standard configuration used in all tests
pipeline_optimizer = TPOTRegressor(
    generations=20, 
    population_size=20, 
    cv=5, 
    random_state=seed,
    verbose=0
)

In v1.0.0, this is run inside 'if __name__ == "__main__":'
pipeline_optimizer.fit(X_train, y_train)

**Questions**
Is this drastic slowdown expected for small datasets due to the overhead of the Dask backend (introduced in v1.0.0) compared to the older multiprocessing backend?

Has the default config_dict changed in v1.0.0 to prioritize significantly heavier estimators (e.g., more aggressive use of XGBoost or Stacking) compared to v0.12.x, even when XGBoost was present in the older environments?

Is there a recommended configuration to restore the runtime profile of the older versions for benchmarking purposes?

**Comment by perib** [comment]
There are several differences between TPOT1 and TPOT2 which can be contributing to this.

One reason is that the default configuration has changed. 
TPOT2: https://github.com/EpistasisLab/tpot/blob/main/tpot/tpot_estimator/estimator.py
TPOT1: https://github.com/EpistasisLab/tpot/blob/master-archived/tpot/base.py

Some differences to highlight:
1) There are variable changes to default parameters. TPOT2 has a higher timeout for individual model evaluations - 10 minutes in TPOT2 vs 5 minutes in TPOT1. Should also note that by default TPOT2 stops after 60 minutes, rather than a set number of generations. you have to manually set this to None to disable it. TPOT2 by default has a smaller population size to encourage more evaluations and less random search. TPOT2 does more crossover and fewer mutations compared to TPOT1, etc. More crossover and smaller population size can contribute to larger pipelines, for example.

We wanted to keep TPOTRegressor simple, with limited parameters. the base class, TPOTEstimator has several more parameters https://github.com/EpistasisLab/tpot/blob/main/tpot/tpot_estimator/estimator.py#L71

TPOT1 parameters can be found here: https://github.com/EpistasisLab/tpot/blob/master-archived/tpot/base.py

2) TPOT2 and TPOT have different models/transformers included, with different possible hyperparameter ranges.
TPOT1: https://github.com/EpistasisLab/tpot/blob/master-archived/tpot/config/regressor.py

TPOT2 config is broken into a few different files you can find here: https://github.com/EpistasisLab/tpot/blob/main/tpot/config
The regressor configuration is defined with this function call here: https://github.com/EpistasisLab/tpot/blob/main/tpot/tpot_estimator/templates/tpottemplates.py#L275

Tutorial 2 explains how to set up search spaces: https://github.com/EpistasisLab/tpot/blob/main/Tutorial/2_Search_Spaces.ipynb

3) The shape of possible pipelines are also different in TPOT2. TPOT1 only supports a tree like shape, TPOT2 defaults to linear* pipelines, but also support graphs, trees, or any flexible configuration with custom restrictions (see the tutorials for examples.) Note that the default TPOT2 pipeline search space is what we found to be the best performing (better than unrestricted graphs or trees). 

The equivalent search space to TPOT1 would be approximately something like:
https://github.com/EpistasisLab/tpot/blob/main/tpot/search_spaces/pipelines/tree.py#L64C7-L64C19

```
TreePipeline(root_search_space <- regressors
                        leaf_search_space <- regressors, transformers, and selectors
                        inner_search_space <- regressors, transformers, and selectors
)
```

* linear pipelines in TPOT2 include middle sections with arbitrary numbers of transfomers/classifiers. E.g Selector -> (transformer1, transformer2, in one layer, each receiving the same input) -> regressor

but you have to create a list of nodes that correspond to the same ranges as in TPOT1


4) by default, TPOT1 also tries to minimize pipeline size, TPOT2 does not (this should probably be changed. ). I recommend using the ,tpot.objectives.complexity_scorer as that is what I have had the best results with. The equivalent to TPOT 1 is tpot.objectives.number_of_leaves_objective. 

for example:
```

scorers=['roc_auc_ovr',tpot.objectives.complexity_scorer],
scorers_weights=[1,-1],
```

The equivalent to TPOT1 would be:
```
scorers=['roc_auc_ovr'],
scorers_weights=[1],
other_objective_functions=[tpot.objectives.number_of_leaves_objective],
other_objective_functions_weights=[-1],
```

In my testing, I found that the best performance would actually be to include both (in some cases, YMMV):
```
scorers=['roc_auc_ovr',tpot.objectives.complexity_scorer],
scorers_weights=[1,-1],
other_objective_functions=[tpot.objectives.number_of_leaves_objective],
other_objective_functions_weights=[-1],
```

5) One more step you can take is to look at the pipelines that are searched by both TPOT1 and TPOT2. In TPOT2, you can find that list here: est.evaluated_individuals

6) One more option for speeding up TPOT runs can be found in this tutorial: https://github.com/EpistasisLab/tpot/blob/main/Tutorial/8_SH_and_cv_early_pruning.ipynb . It would still require some tinkering and would likely mainly work when you have larger datasets, or significantly larger population sizes compared to number of threads. 

More information in the tutorials, which I highly recommend. There are major changes from the previous version, and significantly more customization options: https://github.com/EpistasisLab/tpot/blob/main/Tutorial/
You can find the last version of TPOT1 here: https://github.com/EpistasisLab/tpot/tree/master-archived

**Comment by perib** [comment]
My guess is that dask is heavier than joblib for parallelization, but I wouldn't expect such a large increase. My guess is that the timeout + different configuration space is the largest contributing factor here.


Let me know if you have any questions!

**Comment by christofplie** [comment]
@perib Thank you very much for the quick and detailed answers and suggestions. I have one more question about the **if __name__=="__main__":** wrapper. In this tutorial: https://github.com/EpistasisLab/tpot/blob/main/Tutorial/7_dask_parallelization.ipynb it is mentioned that this wrapper has to be used in order to protect the code because how parallelization is handled in Python. But in my experiment I use an already existing Python script which was tested on older tpot versions (before 1.0.0) and this code does not have this wrapper and it still works. But when I tried using TPOT 1.0.0 the code crashed until i started using the wrapper. Is this now happening because of how Dask is used in new TPOT versions? Because the tutorial mentions that the wrapper is used because of parallelization in Python but then it should apply to my script with older TPOT versions as well.

**Comment by perib** [comment]
While older versions of TPOT used joblib by default, the latest version (1.0.0) uses Dask and no longer supports joblib. You now need the if __name__ == "__main__": block because it is required for Dask's parallelization method, whereas the previous joblib implementation did not require it.

Basically, it comes down to how parallelization works, specifically whether it is done through threads or processes. In dask you can select which option to use. By default, TPOT uses processes. You can change this with the 'processes' parameter. If set to false, you don't need `"if __name__ == "__main__" " `IIRC.

> processes : bool, default=True
>             If True, will use multiprocessing to parallelize the optimization process. If False, will use threading.
>             True seems to perform better. However, False is required for interactive debugging.
> 

Essentially what happens is that with multiprocessing, each worker is completely independent (doesn't share memory) and will execute the .py independently. Without the `"if __name__ == "__main__" ",` the entire script including the TPOT .fit command is executed again. Your script crashes because it eventually uses up all the resources as it calls exponentially more fit commands.  Otherwise, it just loads the required modules.


Another difference in this new version is also the 'scatter' parameter which will copy the data into each worker separately, to avoid having to continuously copy the dataset between workers. 

> 
> scatter : bool, default=True
>             If True, will scatter the data to the dask workers. If False, will not scatter the data. This can be useful for debugging.

**Comment by perib** [comment]
It executes the .py script as part of its attempt to load all modules.
